### ISSUE TITLE
No way to use toolchains correctly in executable rule output

### ISSUE THREAD

**Issue by anguslees** [bazelbuild/bazel#19645]
### Description of the feature request:

A common pattern[^1] I see in rules is toolchain declarations that effectively capture a tool _path_, and templated executable/test rules that resolve the toolchain and substitute the tool path into the executable file output.

Unfortunately this is incorrect!  When the exec and target platforms differ, the toolchain will be resolved for the _exec_ platform, but the executable/test output needs to be for the _target_ platform.

As far as I can see, there is no way to make toolchain resolution / platform transitions do the correct thing here. Since executable rules are how bazel tests work, this is a pretty major gap in the toolchains/platforms story.   I can't file bugs against rules, because there is no reasonable solution to suggest.  This is a feature request for a solution that is (a) correct, and also ideally (b) easy/obvious.

One onerous current workaround is to declare _two_ toolchains for all such cases.  One that uses exec_compatible_with/target_compatible_with as usual, and is used when the toolchain is invoked by an action directly.  And a second toolchain that uses target_compatible_with _only_, which can be used in executable rule output.  This is moderately terrible, since this duplicates work for everybody, confuses the meaning of exec_compatible_with vs target_compatible_with, and toolchain authors can't anticipate when their toolchain needs to be used in this way - undoing the loose coupling between author and consumer that is kind of the point of toolchains.

An incorrect (afaics) version of the two-toolchains approach is to _only_ declare target_compatible_with toolchain, then create a "current_foo_toolchain" rule that resolves the toolchain for a particular target, and returns the toolchain provider through a rule attribute - see eg [python toolchain](https://github.com/bazelbuild/rules_python/blob/main/docs/python.md#current_py_toolchain).  This is incorrect in general because another toolchain that depends on this toolchain cannot be resolved at the same time for the same exec+target combination (ie: `toolchains=[foo]` with a dependency on another rule with `toolchains=[bar]`, is not the same as `toolchains=[foo, bar]`).

As a possible implementation, I'd _love_ to be able to do `exec_group(cfg='target')`, since I think `ctx.exec_groups["target"].toolchains` is an intuitive and 'obvious' way to address the above.

Another option is to _not_ use toolchains for what are effectively artifacts.  Rule authors obviously like the loose coupling between toolchain declaration and use however, so this potential path would require widespread advocacy/documentation, and perhaps a canonical meta-library for an "artifact toolchain".

See also slack thread: https://bazelbuild.slack.com/archives/CA31HN1T3/p1690176577746239

[^1]: Examples of (incorrect) toolchains used for executable rules in [rules_js](https://github.com/aspect-build/rules_js/issues/1168), [rules_docker](https://github.com/bazelbuild/rules_docker/blob/8e70c6bcb584a15a8fd061ea489b933c0ff344ca/contrib/push-all.bzl#L56), [rules_oci](https://github.com/bazel-contrib/rules_oci/blob/54d04c20c6bfa0aa1089385e5abe35af462f0bd7/oci/private/push.bzl#L102-L103), [gazelle](https://github.com/bazelbuild/bazel-gazelle/blob/c11d7ce0133088d69e5a18541f219490cc3c1996/def.bzl#L93)

### Which category does this issue belong to?

Core

### What underlying problem are you trying to solve with this feature?

It should be possible to resolve and refer to the correct toolchain for an executable rule output.

### Which operating system are you running Bazel on?

macos and Linux

### What is the output of `bazel info release`?

6.3.2

**Comment by moroten** [comment]
Any updates on this?

**Comment by katre** [comment]
I'm very confused about your use case, hopefully we can figure out what makes sense.

Typically, the intent of toolchains is that they are used during build actions: a c++ compiler is called, an analyzer is run, whatever.

It sounds like you want some files from a toolchain, but you want your build rule to bundle those into the outputs? We've used toolchains for that in a few cases (the Python runtime is probably the best example), but it's always been a tricky fit, because it's using the tools as an output to be passed along.

Maybe you don't want a toolchain, you want a direct (target configured) dependency on the tools, so that they can be bundled into your rule's output?

**Comment by anguslees** [comment]
> Maybe you don't want a toolchain, you want a direct (target configured) dependency on the tools, so that they can be bundled into your rule's output?

I agree, but see my numerous linked examples in the footnote of the opening post.  This mis-pattern is quite widespread, because folks (apparently) _want_ to use toolchains in places like tests or bazel-run scripts (both executable rules).

This is a feature request for an improvement to bazel toolchains/exec_groups to make this possible.  Or alternatively we loudly declare this will 'never' be possible, and (as a community), we go and rewrite these examples to avoid toolchains.

**Comment by katre** [comment]
Toolchains (uniquely) have two options for their dependencies:
- Anything with `cfg = "exec"` is configured for the exec platform, which is the same as the requesting target's exec platform
- Anything with `cfg = "target"` (which is the default if `cfg` isn't specified) is in the same configuration as the requesting target.

So part of the solution is probably to be more careful of the target/exec distinction, here. It's unfortunately easy to get this wrong in the default case, since the base case is that the host platform is used as both the only exec platform and as the default target platform.

I agree that there is probably more we can be doing to make this cleaner. I'm going to add this item to our backlog for tracking purposes. If you have any specific features you'd like to request, we'd be happy to see a proposal or further discussion to move this along.

**Comment by anguslees** [comment]
> So part of the solution is probably to be more careful of the target/exec distinction, here.

I think we're talking past each other.  I (believe I am) familiar with cfg=target/cfg=exec and when to use them.  This does not help with the problem described in this issue.

I'll explain one example here in detail, to see if that helps:

rules_oci has a target that can be used similar to `bazel run //my/image:push` to upload an artifact (it's an OCI image, but the specific file format is not relevant) to a repository.   As with all other bazel-run targets, the rule for this `//my/image:push` target is [an "executable" rule](https://bazel.build/extending/rules#executable_rules_and_test_rules), which returns the file (commonly a script) that is _going_ to be executed, but doesn't actually run the file itself.

This executable rule in this case is the following:

```starlark
def _impl(ctx):
    crane = ctx.toolchains["@rules_oci//oci:crane_toolchain_type"]
    yq = ctx.toolchains["@aspect_bazel_lib//lib:yq_toolchain_type"]

    if not ctx.file.image.is_directory:
        fail("image attribute must be a oci_image or oci_image_index")

    _, _, _, maybe_digest, maybe_tag = util.parse_image(ctx.attr.repository)
    if maybe_digest or maybe_tag:
        fail("`repository` attribute should not contain digest or tag. got: {}".format(ctx.attr.repository))

    executable = ctx.actions.declare_file("push_%s.sh" % ctx.label.name)
    files = [ctx.file.image]
    substitutions = {
        "{{crane_path}}": crane.crane_info.binary.short_path,
        "{{yq_path}}": yq.yqinfo.bin.short_path,
        "{{image_dir}}": ctx.file.image.short_path,
        "{{fixed_args}}": " ".join(_quote_args(["--repository", ctx.attr.repository])),
    }
    if ctx.attr.remote_tags:
        files.append(ctx.file.remote_tags)
        substitutions["{{tags}}"] = ctx.file.remote_tags.short_path

    ctx.actions.expand_template(
        template = ctx.file._push_sh_tpl,
        output = executable,
        is_executable = True,
        substitutions = substitutions,
    )
    runfiles = ctx.runfiles(files = files)
    runfiles = runfiles.merge(yq.default.default_runfiles)
    runfiles = runfiles.merge(crane.default.default_runfiles)

    return DefaultInfo(executable = executable, runfiles = runfiles)

# Refactored slightly from the original for clarity
oci_push = rule(
    implementation = _impl,
    attrs = _attrs,
    toolchains = [
        "@rules_oci//oci:crane_toolchain_type",
        "@aspect_bazel_lib//lib:yq_toolchain_type",
    ],
    executable = True,
)
```

To summarise the relevant parts: this rule gets the paths for two tools (`crane` and `yq`) using bazel toolchains, substitutes those paths into a shell script template, and then returns that shell script as the executable output file that will eventually be invoked by bazel-run.

But this use of toolchains is incorrect when cross-compiling!

Specifically: When bazel host platform != exec platform (eg: run bazel on macos with a linux RBE cluster), the rule implementation above will run on the exec (linux) platform, but the output executable needs to work on the host (macos) platform.  The correct `crane` and `yq` tools to use here, are the tools **for the target** platform of this rule, and not the exec platform.

**But there is NO WAY to resolve the toolchain for the target of a rule!**

^^This is the bazel limitation described in this github issue.   If there is anything I can do to expand or clarify the above explanation, please suggest it because it seems to be subtle/difficult for folks to recognise the issue here and I'd like to point folks to this comment.

Afaik cfg=target is not relevant here, since there is nowhere to put that that will affect toolchain resolution - the `ctx.toolchains["@rules_oci//oci:crane_toolchain_type"]` result will always be resolved for the exec platform.

Afaics this incorrect use (or misunderstanding) of toolchains is moderately common (see my examples in widely used rules), and can impact both bazel-run and bazel tests (both executable rules).  Ideally, we improve bazel so that toolchains _can_ be used in these important use cases.

**Comment by moroten** [comment]
This looks very related to #21805, where I suggest the `--run_under` target should be built for the host platform.

**Comment by anguslees** [comment]
> This looks very related to https://github.com/bazelbuild/bazel/discussions/21805

Related, but not the same - just to prevent an overzealous close-as-duplicate.  I agree `--run_under` (and `run` generally) should be build for host platform.  Providing a better default --platform won't fix this no-way-to-resolve-toolchains-for-non-exec-platform issue.

**Comment by EdSchouten** [comment]
Hey @anguslees,

I did some experimenting, and it looks like what you want is possible by doing the following:

1. You move the toolchains into a helper rule.
2. You instantiate a target of that helper rule in some BUILD file.
3. You let the original rule depend on the target.
4. You apply an outgoing edge transition on that dependency.

The outgoing edge dependency can make use of the following transition:

```python
def _transition_exec_to_target_impl(settings, attr):
     return {
         "//command_line_option:extra_execution_platforms": settings["//command_line_option:platforms"],
     }

 transition_exec_to_target = transition(
     implementation = _transition_exec_to_target_impl,
     inputs = ["//command_line_option:platforms"],
     outputs = ["//command_line_option:extra_execution_platforms"],
 )
```

I will send out a PR against rules_oci soon to get `oci_push()` fixed.

That said, I do think that there is a bit of voodoo magic going on. Transitioning on command line flags is a bit funky. Especially because we're doing it on `--extra_execution_platforms` (which is fortunately documented as prepending to the list of registered platforms, which is exactly what we want). Maybe it makes sense to add something like this, either to `@bazel_tools`, `@bazel_skylib`, or `@aspect_bazel_lib`. Any preference for a place?

**Comment by anguslees** [comment]
Oh wow - I remember considering this, but I think I must have just incorrectly concluded that transitioning `--extra_execution_platforms` wouldn't work, without trying.

Fwiw on bazel 6.4.0, I needed to force //command_line_option:platforms to a string first:

```starlark
def _transition_exec_to_target_impl(settings, attr):
     return {
         "//command_line_option:extra_execution_platforms": str(settings["//command_line_option:platforms"][0]),
     }
```

Combined with [the `resolved_toolchain()` idiom](https://github.com/bazel-contrib/rules_oci/blob/a86249d4640b00118796977a3fa6e68329d9a7d4/oci/private/toolchains_repo.bzl#L83-L93) I've seen around a few places, this is actually not terrible.  In fact I suspect many folks who have copied `resolved_toolchain` from java - but are _not_ actually portable across platforms - need to add this transition in order to be correct.

**Comment by katre** [comment]
Instead of using a transition to add new execution platforms, consider using a [flag to activate them](https://github.com/bazelbuild/proposals/blob/main/designs/2024-04-26-platform-exec-settings.md) (once I sit down and implement the proposal, hopefully this week).

**Comment by EdSchouten** [comment]
Hey @katre,

Even though that proposal you linked looks interesting, I don’t really understand how it’s relevant in this context. In this specific case we want to force toolchain resolution so that the resulting toolchain can be executed on whatever the current target platform is. I don’t think that can be solved by conditionally enabling registration of platforms.

**Comment by EdSchouten** [comment]
> This can be solved by defining two toolchain_type for the toolchain. `toolchain_type(name="tool")` and `toolchain_type(name="runtime")`. We already do this for Python and Java.

Are those the toolchains that intentionally have exec/target_compatible_with set the wrong way around? I’ve never been a fan of that approach.

**Comment by EdSchouten** [comment]
> This can be solved by defining two toolchain_type for the toolchain. `toolchain_type(name="tool")` and `toolchain_type(name="runtime")`. We already do this for Python and Java.

Are those the toolchains that intentionally have exec/target_compatible_with set the wrong way around? I’ve never been a fan of that approach.

**Comment by fmeum** [comment]
> Are those the toolchains that intentionally have exec/target_compatible_with set the wrong way around? I’ve never been a fan of that approach.

I don't know how Python does this, but I'm pretty sure that the Java setup is correct:
https://github.com/bazelbuild/bazel/blob/2fb8564f7b0abfb669685df4873ffb8a5bb14741/tools/jdk/BUILD.tools#L11

**Comment by anguslees** [comment]
> This can be solved by defining two toolchain_type for the toolchain

Yes.  I tried to respond to this solution in my original issue description:

> One onerous current workaround is to declare two toolchains for all such cases. One that uses exec_compatible_with/target_compatible_with as usual, and is used when the toolchain is invoked by an action directly. And a second toolchain that uses target_compatible_with only, which can be used in executable rule output. This is moderately terrible, since this duplicates work for everybody, confuses the meaning of exec_compatible_with vs target_compatible_with, and toolchain authors can't anticipate when their toolchain needs to be used in this way - undoing the loose coupling between author and consumer that is kind of the point of toolchains.

I think those objections still stand.  You're basically suggesting that _every_ "tool" toolchain should declare/register two toolchains, since the toolchain providers can't anticipate whether a consumer is going to want it as an "exec" or a "target" toolchain.  I agree that works; in composing this feature request, I felt this was unacceptably onerous on the entire ecosystem - or at least requires a "decision" that this is how we want to use toolchains, that we can point to and rally behind.

Note: An aside for others casually skimming this discussion:
- rules_python registers a "target only" toolchain (no exec constraints), and uses it through a "[current_py_toolchain](https://github.com/bazelbuild/rules_python/blob/main/python/current_py_toolchain.bzl)" dependency - essentially turning the toolchain into a dep.  This works, but moving exec constraints to target constraints in this way is (I claim) confusing, and fails when toolchains depend on other toolchains (eg python might require cc toolchain), since it loses the ability to represent separate exec/target constraints when required.
- rules_java registers _two_ toolchains: one intended for use at execution time (exec constraints) and one for runtime (target constraints).  This seems appropriate for Java, which has a historical separation between execution (jdk) and runtime (jre), and of course has standardised cross-platform artifacts.  I don't think it's reasonable to generalise this to other languages/toolchains.

**Comment by anguslees** [comment]
> Not very toolchain, only those who want to be used in cfg=target

But the toolchain _author_ couldn't know if their toolchain was going to be used in a test/run - so yes, we would have to pre-emptively declare and configure every toolchain twice, just in case.

**Comment by lberki** [comment]
Let me see if I understand this correctly: in the current model of Bazel, a toolchain has two configurations: the one it runs on and the one it targets (e.g. a Rust compiler can run on x64 and produce ARM binaries). So each toolchain has to be tagged with these two configurations and Bazel selects one that matches the exec and target configurations and it assumes that the reason why one uses a toolchain is to run it during the build and consume its output.

However, some people want to use the loose coupling afforded by toolchains to choose what to incorporate into their actual outputs: say, one may want to produce an SDK that runs on x64 and package an x64 to ARM compiler into it. In this case, the right thing to do is to match the "target" configuration of Bazel to the "exec" configuration of the toolchain. Bazel currently can't to that. 

The "two toolchain" workaround is to wrap every toolchain in two toolchain rules: one that's "honest" and one that "lies", in the above case saying that the target configuration of the compiler is x64. This is obviously wrong because, well, that's not what the toolchain does.

@katre any clever ideas here? I can imagine a lot of things but you have a lot more state about this in your head than I do.

**Comment by lberki** [comment]
For Java, that's arguably the case since the JDK and the JVM are different, but what about a C++ compiler or just a regular tool? If you accept the above SDK use case as valid, Bazel currently forces you to put the exec configuration of the C++ toolchain into its `target_compatible_with` attribute for Bazel to do the right thing.

I think the conceptual difference between us is that you consider "toolchain to be used in the build" and "toolchain to be packaged as the result of the build" different things; in my mind, they are just one "thing" but two different use cases: it's true that one needs to provide much more rich information about the toolchain to be invoked during the build than just to package it, but a list of files the toolchain is comprised of is necessary for both and it feels silly to duplicate the exec platform compatibility information and the set of files like this.

It's a reasonable workaround, but it requires the owner of every toolchain to cooperate and very public interfaces should be very narrow.

**Comment by fmeum** [comment]
I agree with @BoleynSu here. The current approach is conceptually sound, the problem is that it's apparently too verbose or too hard to use correctly for simple cases that don't need the full flexibility afforded by toolchains. 

That's where the `resolved_X_binary` rule mentioned in https://github.com/bazelbuild/bazel/issues/19645#issuecomment-2227266641 can come in to help. Maybe we just need to make this pattern less verbose by abstracting it into a shared helper macro?

**Comment by lberki** [comment]
`resolved_*_binary()` isn't a full solution, is it? In particular, it allows one to decide what toolchain to use based on either the exec platform or the target platform, but not both (i.e. it doesn't support the "this compiler runs on x64 and emits ARM code"). So `resolved_*_binary()` helps, but it needs to be used in addition to the "regular" toolchain things.

I don't think Bazel allows one to match the constraints of the target platform to the exec constraints of a toolchian, does it? (except by the awkward transition above https://github.com/bazelbuild/bazel/issues/19645#issuecomment-2127875342 )

There is some merit in the argument that "toolchain as functionality" and "toolchain as bunch of files" are different, though. The awkwardness here is that if it's to work, every author of a toolchain for a particular type must cooperate and it's extra boilerplate for all of them.

**Comment by EdSchouten** [comment]
> `resolved_*_binary()` isn't a full solution, is it? In particular, it allows one to decide what toolchain to use based on either the exec platform or the target platform, but not both (i.e. it doesn't support the "this compiler runs on x64 and emits ARM code"). So `resolved_*_binary()` helps, but it needs to be used in addition to the "regular" toolchain things.

Exactly. This is also the reason I've been highly skeptical of that approach.

**Comment by katre** [comment]
This is a pretty interesting discussion, and the distinction between "toolchain which provides a binary as a tool for actions" and "toolchain which provides a binary to be packaged in an artifact" is a pretty subtle one.

I think there are several patterns in this thread that could be applied to rules, using the current APIs. So, I think the next step is to figure out what is the correct outcome of this issue?

Idea 1: Someone (possibly @BoleynSu?) could write up some documentation for bazel.build describing when this pattern is useful, and what rule authors need to do to use it.

Idea 2: Possibly simultaneously, someone (@BoleynSu? @fmeum?) could write some helper macros for constructing these distinct varieties of toolchains (updating the documentation from the step above).

Question 1: Is there an actual request for changes in Bazel (presumably in toolchain resolution) here? 

From the discussion so far, this feels like a problem in 1) documenting the underlying patterns and when they are useful, and 2) convincing rule authors to use them to make these available to users. Is that a reasonable summary?

**Comment by lberki** [comment]
The only goal I have here is that this use case is doable somehow. I don't have a preference towards changing Bazel or providing libraries to do so in principle, however, in this case, the former seems more reasonable because (in order of what I think is the strongest to the weakest argument): 
1. In order for a toolchain type to be usable this way, all the toolchains that implement it would have to use whatever wrapper we'd cook up
2. The necessary information is already there in Bazel, but not usable by users 
3. I make sense conceptually to represent a toolchain as one "thing" and all proposed wrapper code breaks this

**Comment by lberki** [comment]
I'm not sure "making Bazel less complex" is the right thing to optimze for here.

It's a good idea to optimize for "less complexity is seen by the people who use Bazel", but that complexity is not just that of Bazel, but also that of all external Bazel code they depend on and that includes toolchain type definitions and their implementations. IOW, if we can buy simplicity in many toolchains with a little additional wrinkle in Bazel, that's a good deal.

That said, your argument that toolchains are not necessarily single groups of files is a good one for arguing *against* finding a solution in Bazel (for example, even the C++ compiler is not just a single group of file because Bazel has the ability to take different sets of inputs for e.g. compile and link actions).

**Comment by katre** [comment]
Coming back to this: @lberki, I can see your point, but I'm very unclear if there's an actual request besides "make this easier".

I'm happy to review and discuss a concrete proposal, but I'm not sure the Configurability team can actually spend any time on this in the near future.

**Comment by lberki** [comment]
@katre fair enough; I think the insight above that a toolchain doesn't even necessarily need to provide any files to be a toolchain (which manifests in the code as toolchains being free to provider any provider, `DefaultInfo` is not necessarily needed) makes it undesirable that we provide a general "put the files comprising this toolchain on the inputs of that action" functionality.

I think what may make sense is being able to ask Bazel "give me a toolchain whose host configuration matches my target configuration", but given that it has a workaround in Starlark (declaring two toolchains and two toolchain types) and that we'd have to decide what target configuration that toolchain should have, I don't think we should treat this as a priority.

I guess if there was a concrete proposal to do so, @katre would entertain it, wouldn't he?

**Comment by archshift** [comment]
This is so important!

**Comment by katre** [comment]
@archshift Did anyone ever submit a concrete proposal for what needs to be changed? I'm still happy to review such a thing and make recommendations to the Bazel team.

**Comment by BoleynSu** [comment]
Let me draft one but it may take some time.
